### ISSUE TITLE
port etcd joint consensus

### ISSUE THREAD

**Issue by BusyJay** [tikv/raft-rs#378]
We have developed ours implementation but as discussed in the past, we decided to go with the community and port the implement of etcd instead.

There will be several PRs:

- [ ] https://github.com/etcd-io/etcd/pull/10779
- [ ] https://github.com/etcd-io/etcd/pull/10865
- [ ] https://github.com/etcd-io/etcd/pull/10884
- [ ] https://github.com/etcd-io/etcd/pull/10889
- [ ] https://github.com/etcd-io/etcd/pull/10914
- [ ] https://github.com/etcd-io/etcd/pull/11003
- [ ] https://github.com/etcd-io/etcd/pull/11005
- [ ] https://github.com/etcd-io/etcd/pull/11046

The PRs doesn't have to be picked one by one. And we probably sync the code with upstream first.

**Comment by Fullstop000** [comment]
Does #318 need to be merged first?

**Comment by BusyJay** [comment]
I will send a new PR and do it in a different way.

**Comment by accelsao** [comment]
I would like to try #10779

**Comment by BusyJay** [comment]
Thanks @accelsao! I'm already working on it.

**Comment by BusyJay** [comment]
@accelsao Hi, are you interested in porting data driven tests? I have tried to contact you on slack, but it seems you don't check it recently.

**Comment by accelsao** [comment]
ok, I will try it

**Comment by BusyJay** [comment]
Thanks! You can do it package by package. You can also contact me on slack or github if you have any questions.

**Comment by drogus** [comment]
Are there any updates on these issues? I'd be happy to help if there's still something left to do.

**Comment by accelsao** [comment]
only tests are left
1. majority-joint-commit
2. vote
3. group-commit
